source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -0,0 +1,267 @@
+ProcessCluster(const NetworkFilters::Common::Redis::RespValue& value) {
+  ASSERT(value.type() == NetworkFilters::Common::Redis::RespType::Array);
+  ASSERT(value.asArray().size() >= 2);
+  std::string address = value.asArray()[0].asString();
+  bool ipv6 = (address.find(":") != std::string::npos);
+  if (ipv6) {
+    return std::make_shared<Network::Address::Ipv6Instance>(address,
+                                                            value.asArray()[1].asInteger());
+  }
+  return std::make_shared<Network::Address::Ipv4Instance>(address, value.asArray()[1].asInteger());
+}
+
+RedisCluster::RedisDiscoverySession::~RedisDiscoverySession() {
+  if (current_request_) {
+    current_request_->cancel();
+    current_request_ = nullptr;
+  }
+
+  if (client_) {
+    client_->close();
+  }
+}
+
+void RedisCluster::RedisDiscoverySession::onEvent(Network::ConnectionEvent event) {
+  if (event == Network::ConnectionEvent::RemoteClose ||
+      event == Network::ConnectionEvent::LocalClose) {
+    // This should only happen after any active requests have been failed/cancelled.
+    ASSERT(!current_request_);
+    parent_.dispatcher_.deferredDelete(std::move(client_));
+  }
+}
+
+void RedisCluster::RedisDiscoverySession::registerDiscoveryAddress(
+    const std::list<Envoy::Network::Address::InstanceConstSharedPtr>& address_list) {
+  discovery_address_list_.insert(discovery_address_list_.end(), address_list.begin(),
+                                 address_list.end());
+}
+
+void RedisCluster::RedisDiscoverySession::startResolve() {
+  parent_.info_->stats().update_attempt_.inc();
+  // if a resolution is current in progress, skip
+  if (current_request_) {
+    return;
+  }
+
+  // if hosts is empty, we haven't got a successful result from CLUSTER SLOTS call yet, pick a
+  // random discovery address from dns
+  Upstream::HostSharedPtr host;
+  if (hosts_.empty()) {
+    const int rand_idx = parent_.random_.random() % discovery_address_list_.size();
+    auto it = discovery_address_list_.begin();
+    std::next(it, rand_idx);
+    host = Upstream::HostSharedPtr{new RedisHost(parent_.info(), "", *it, parent_, true)};
+  } else {
+    const int rand_idx = parent_.random_.random() % hosts_.size();
+    host = hosts_[rand_idx];
+  }
+
+  if (!client_) {
+    client_ = client_factory_.create(host, parent_.dispatcher_, *this);
+    client_->addConnectionCallbacks(*this);
+  }
+
+  current_request_ = client_->makeRequest(ClusterSlotsRequest::instance_, *this);
+}
+
+void RedisCluster::RedisDiscoverySession::onResponse(
+    NetworkFilters::Common::Redis::RespValuePtr&& value) {
+  current_request_ = nullptr;
+  // TODO(hyang): move to use connection pool for the cluster and avoid closing the connection each
+  // time
+  if (client_) {
+    client_->close();
+  }
+
+  if (value->type() != NetworkFilters::Common::Redis::RespType::Array) {
+    ENVOY_LOG(error, "Unexpected response to cluster slot command: {}", value->toString());
```

Comment:
I would probably do warning here and not error.

Reply:
ok